Core/PersistentPixelStoring.swift

```diff
@@ -0,0 +1,157 @@
+    enum PixelType: Codable {
+        case daily
+        case count
+        case regular
+    }
+
+    let eventName: String
+    let pixelType: PixelType
+    let additionalParameters: [String: String]
+    let includedParameters: [Pixel.QueryParameters]
+
+    var pixelName: String {
+        switch pixelType {
+        case .daily: return eventName + "_d"
+        case .count: return eventName + "_c"
+        case .regular: return eventName
+        }
+    }
+
+    var timestamp: String? {
+        return additionalParameters[PixelParameters.originalPixelTimestamp]
+    }
+}
+
+protocol PersistentPixelStoring {
+    func append(pixel: PersistentPixelMetadata) throws
+    func replaceStoredPixels(with pixels: [PersistentPixelMetadata]) throws
+    func storedPixels() throws -> [PersistentPixelMetadata]
+}
+
+public enum PersistentPixelStorageError: Error {
+    case readError(Error)
+    case writeError(Error)
+    case encodingError(Error)
+    case decodingError(Error)
+}
+
+final class DefaultPersistentPixelStorage: PersistentPixelStoring {
+
+    enum Constants {
+        static let queuedPixelsFileName = "queued-pixels.json"
+        static let pixelCountLimit = 100
+    }
+
+    private let fileManager: FileManager
+    private let fileName: String
+    private let storageDirectory: URL
+    private let pixelCountLimit: Int
+
+    private let fileAccessQueue = DispatchQueue(label: "Persistent Pixel File Access Queue", qos: .utility)
+    private let encoder = JSONEncoder()
+    private let decoder = JSONDecoder()
+
+    private var fileURL: URL {
+        return storageDirectory.appendingPathComponent(fileName)
+    }
+
+    init(fileManager: FileManager = .default,
+         fileName: String = Constants.queuedPixelsFileName,
+         storageDirectory: URL? = nil,
+         pixelCountLimit: Int = Constants.pixelCountLimit) {
+        self.fileManager = fileManager
+        self.fileName = fileName
+        self.pixelCountLimit = pixelCountLimit
+
+        if let storageDirectory = storageDirectory {
+            self.storageDirectory = storageDirectory
+        } else if let appSupportDirectory = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first {
+            self.storageDirectory = appSupportDirectory
+        } else {
+            fatalError("Unable to locate application support directory")
+        }
+    }
+
+    func append(pixel: PersistentPixelMetadata) throws {
+        try fileAccessQueue.sync {
+            var pixels = try self.readStoredPixelDataFromFileSystem()
+            pixels.append(pixel)
+
+            if pixels.count > pixelCountLimit {
+                pixels.removeFirst()
+            }
+
+            try writePixelDataToFileSystem(pixels: pixels)
+        }
+    }
+
+    func replaceStoredPixels(with pixels: [PersistentPixelMetadata]) throws {
+        try fileAccessQueue.sync {
+            try writePixelDataToFileSystem(pixels: pixels)
```

Comment:
Shouldn't we also check the `pixelCountLimit` here?

Reply:
Done; this function has been replaced with new `append` and `remove` functions, where the `append` one can append an array of any size but will take the last 100 elements before writing them back to disk.